api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+}
+
+type Service struct {
+	clientBuilder lotus.ClientBuilder
+	col           *mongo.Collection
+	server        *grpc.Server
+	waiting       map[primitive.ObjectID]chan waitResult
+	config        Config
+	ticker        *time.Ticker
+	waitingLck    sync.Mutex
+}
+
+type Config struct {
+	Listener           net.Listener
+	ClientBuilder      lotus.ClientBuilder
+	MongoUri           string
+	MongoDbName        string
+	MessageWaitTimeout time.Duration
+	MessageConfidence  uint64
+	RetryWaitFrequency time.Duration
+	Debug              bool
+}
+
+func New(ctx context.Context, config Config) (*Service, error) {
+	if config.Debug {
+		if err := util.SetLogLevels(map[string]logging.LogLevel{
+			"sendfil": logging.LevelDebug,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.MongoUri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.MongoDbName)
+	col := db.Collection(collectionName)
+	if _, err := col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys: bson.D{primitive.E{Key: "from", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "to", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "amount_nano_fil", Value: 1}},
+		},
+		// MongoDB automatically creates a multikey index if any indexed field is an array;
+		// you do not need to explicitly specify the multikey type.
+		// https://docs.mongodb.com/manual/core/index-multikey/
+		{
+			Keys:    bson.D{primitive.E{Key: "message_cids.cid", Value: 1}},
+			Options: options.Index().SetUnique(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "message_state", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "waiting", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "updated_at", Value: 1}},
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("creating collection indexes: %v", err)
+	}
+
+	s := &Service{
+		clientBuilder: config.ClientBuilder,
+		col:           col,
+		waiting:       make(map[primitive.ObjectID]chan waitResult),
+		config:        config,
+		ticker:        time.NewTicker(config.RetryWaitFrequency),
+	}
+
+	s.server = grpc.NewServer()
+	go func() {
+		pb.RegisterSendFilServiceServer(s.server, s)
+		if err := s.server.Serve(config.Listener); err != nil && !errors.Is(err, grpc.ErrServerStopped) {
+			log.Errorf("serve error: %v", err)
+		}
+	}()
+
+	if err := s.waitAllQualifying(ctx, true); err != nil {
+		return nil, fmt.Errorf("calling waitAllQualifying: %v", err)
+	}
+
+	s.bindTicker(ctx)
+
+	return s, nil
+}
+
+func (s *Service) waitAllQualifying(ctx context.Context, isInitialRun bool) error {
+	filter := bson.M{"$and": bson.A{
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_ACTIVE}},
+		bson.M{"message_state": bson.M{"$ne": pb.MessageState_MESSAGE_STATE_FAILED}},
+	}}
+	if !isInitialRun {
+		filter["waiting"] = false
+	}
```

Comment:
To detect txns that need to be waited on, we always ask for all messages that aren't in a final state. If it is the "init" call of this func, that is all we need to do, but if this is a call driven by the ticker, we also want to exclude any txns that are already in a waiting state.

Reply:
Pending - we have not yet detected the message is active on chain
Waiting - this service has an active call to lotus `StateWaitMsg` running